internal/sourcemap/metadata_fetcher.go

```diff
@@ -0,0 +1,320 @@
+			// already in the cache, remove from the updates.
+			delete(sopurcemaps, id)
+
+			// content hash changed, invalidate the sourcemap cache
+			if contentHash != updatedHash {
+				s.logger.Debugf("Hash changed: %s -> %s: invalidating %v", contentHash, updatedHash, id)
+				invalidation = append(invalidation, id)
+			}
+		} else {
+			// the sourcemap no longer exists in ES.
+			// invalidate the sourcemap cache.
+			invalidation = append(invalidation, id)
+
+			// the sourcemap no longer exists in ES.
+			// remove from metadata cache
+			delete(s.set, id)
+
+			// remove aliases
+			for _, k := range GetAliases(id.name, id.version, id.path) {
+				delete(s.alias, k)
+			}
+		}
+	}
+
+	s.invalidationChan <- invalidation
+
+	// add new sourcemaps to the metadata cache.
+	for id, contentHash := range sopurcemaps {
+		s.set[id] = contentHash
+		s.logger.Debugf("Added metadata id %v", id)
+		// store aliases with a pointer to the original id.
+		// The id is then passed over to the backend fetcher
+		// to minimize the size of the lru cache and
+		// and increase cache hits.
+		for _, k := range GetAliases(id.name, id.version, id.path) {
+			s.logger.Debugf("Added metadata alias %v -> %v", k, id)
+			s.alias[k] = &id
+		}
+	}
+
+	s.logger.Debugf("Metadata cache now has %d entries.", len(s.set))
+}
+
+func (s *MetadataESFetcher) initialSearch(ctx context.Context, updates map[Identifier]string) (*esSearchSourcemapResponse, error) {
+	resp, err := s.runSearchQuery(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to run initial search query: %w", err)
+	}
+	defer resp.Body.Close()
+
+	return s.handleUpdateRequest(resp, updates)
+}
+
+func (s *MetadataESFetcher) runSearchQuery(ctx context.Context) (*esapi.Response, error) {
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(queryMetadata()); err != nil {
+		return nil, fmt.Errorf("failed to encode metadata query: %w", err)
+	}
+
+	req := esapi.SearchRequest{
+		Index:          []string{s.index},
+		Body:           &buf,
+		TrackTotalHits: true,
+		Scroll:         time.Minute,
+	}
+	return req.Do(ctx, s.esClient)
+}
+
+func queryMetadata() map[string]interface{} {
+	return search(
+		sources([]string{"service.*", "file.path", "content_sha256"}),
+	)
+}
+
+type esSearchSourcemapResponse struct {
+	ScrollID string `json:"_scroll_id"`
+	esSourcemapResponse
+}
+
+func (s *MetadataESFetcher) handleUpdateRequest(resp *esapi.Response, updates map[Identifier]string) (*esSearchSourcemapResponse, error) {
+	// handle error response
+	if resp.StatusCode >= http.StatusMultipleChoices {
+		b, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read response body: %w", err)
+		}
+		if resp.StatusCode == http.StatusNotFound || resp.StatusCode == http.StatusForbidden {
+			return nil, fmt.Errorf("%w: %s: %s", ErrFetcherUnvailable, resp.Status(), string(b))
+		}
+		return nil, fmt.Errorf("ES returned unknown status code: %s", resp.Status())
```

Comment:
We should also explicitly handle if ES is not yet reachable.

Reply:
I've pushed some changes, it should correctly handle ES not being available now